### PR TITLE
Use std::isnan instead of self-comparison.

### PR DESCRIPTION
**Pull request by ezyang**
Stack:
* **#18021 Use std::isnan instead of self-comparison.**

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D14460699](https://our.internmc.facebook.com/intern/diff/D14460699)

### CODE REVIEW

File: aten/src/ATen/native/Sorting.cpp [hunk 1]
```diff
@@ -171,7 +171,7 @@ std::tuple<Tensor&, Tensor&> kthvalue_out_cpu(
               tmp_values,
               k - 1,
               [](scalar_t x, scalar_t y) -> bool {
-                return ((x != x && y == y) || (x > y));
+                return ((std::isnan(x) && !std::isnan(y)) || (x > y));
```
Comment (resistor) [inline aten/src/ATen/native/Sorting.cpp]: Does this work correctly and efficiently when scalar_t is an integral type?  I *think* this casts it to a double and still does a NaN check.

Comment (ezyang) [reply aten/src/ATen/native/Sorting.cpp]: Looks like you're right: https://godbolt.org/z/HhgWIu

Comment (resistor) [reply aten/src/ATen/native/Sorting.cpp]: I believe you should be avale to use numeric traits to test if scalar_t is a floating point type.

Comment (ezyang) [reply aten/src/ATen/native/Sorting.cpp]: It looks like someone already wrote a function do this, I'm going to use it.

**Comment by resistor** [comment]
LGTM

**Comment by ezyang** [comment]
But it doesn't work on Windows XD

**Review by t-vi — APPROVED** [review]
Looks good to me, thanks!

**Pull request closed without merging**